File: tcp_examples/bot_client.py
#!/usr/bin/env python3

import socket
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
import curses  # To handle real-time keypresses

class TcpClientNode(Node):
    def __init__(self):
        super().__init__('bot_client_node')

        # Create a TCP client socket
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.connect(('127.0.0.1', 2000))  # Connect to the server at localhost:2000

        # Create a publisher to send messages to a ROS topic (optional)
        self.pub = self.create_publisher(String, 'teleop_status', 10)

        self.get_logger().info("TCP Client Node initialized. Use keys for control: ")
        self.get_logger().info("'i' = Move Forward, 'j' = Turn Left, 'm' = Move Backward, 'l' = Turn Right, 'k' = Stop")

        self.teleoperate()

    def send_command(self, command: str):
        """Send the command to the server"""
        self.sock.sendall(command.encode())

        # Optionally, you can log the sent command to a topic
        msg = String()
        msg.data = f"Sent command: {command}"
        self.pub.publish(msg)

    def teleoperate(self):
        """Handle key-based user input and control the TurtleBot"""
        stdscr = curses.initscr()  # Initialize the curses screen
        curses.cbreak()  # Disable line buffering
        stdscr.keypad(1)  # Enable special key input (e.g., arrows)

        try:
            while True:
                key = stdscr.getch()  # Get the pressed key

                if key == ord('i'):  # Move forward
                    self.send_command("move_forward")
                elif key == ord('j'):  # Turn left
                    self.send_command("turn_left")
                elif key == ord('m'):  # Move backward
                    self.send_command("move_backward")
                elif key == ord('l'):  # Turn right
                    self.send_command("turn_right")
                elif key == ord('k'):  # Stop the robot
                    self.send_command("stop")
                elif key == ord('q'):  # Quit (exit) the program
                    self.get_logger().info("Exiting control loop.")
                    break

        finally:
            curses.endwin()  # Clean up curses

        # Close the connection after quitting
        self.sock.close()

def main(args=None):
    rclpy.init(args=args)
    client = TcpClientNode()

    rclpy.spin(client)

    client.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/bot_client.cpp
#include <iostream>
#include <string>
#include <unistd.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <curses.h>

class TcpClientNode : public rclcpp::Node
{
public:
    TcpClientNode()
        : Node("bot_client_node"), stop_requested(false)
    {
        // Create a TCP client socket
        sockfd = socket(AF_INET, SOCK_STREAM, 0);
        if (sockfd == -1) {
            RCLCPP_ERROR(this->get_logger(), "Socket creation failed!");
            exit(1);
        }

        server_address.sin_family = AF_INET;
        server_address.sin_port = htons(2000); // Server port
        server_address.sin_addr.s_addr = inet_addr("127.0.0.1"); // Server address

        // Connect to the server
        if (connect(sockfd, (struct sockaddr *)&server_address, sizeof(server_address)) < 0) {
            RCLCPP_ERROR(this->get_logger(), "Connection failed!");
            exit(1);
        }

        // Create a publisher to send messages to a ROS topic (optional)
        pub_ = this->create_publisher<std_msgs::msg::String>("teleop_status", 10);

        RCLCPP_INFO(this->get_logger(), "TCP Client Node initialized. Use keys for control: ");
        RCLCPP_INFO(this->get_logger(), "'i' = Move Forward, 'j' = Turn Left, 'm' = Move Backward, 'l' = Turn Right, 'k' = Stop");

        teleoperate();
    }

    ~TcpClientNode()
    {
        // Close the socket
        close(sockfd);
    }

private:
    int sockfd;
    struct sockaddr_in server_address;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
    bool stop_requested;

    void send_command(const std::string &command)
    {
        // Send the command to the server
        send(sockfd, command.c_str(), command.length(), 0);

        // Optionally, log the sent command to a ROS topic
        auto msg = std_msgs::msg::String();
        msg.data = "Sent command: " + command;
        pub_->publish(msg);
    }

    void teleoperate()
    {
        // Initialize curses screen
        initscr();
        cbreak();
        keypad(stdscr, TRUE);
        noecho();

        try {
            while (!stop_requested) {
                int key = getch();  // Get the pressed key

                if (key == 'i') {  // Move forward
                    send_command("move_forward");
                }
                else if (key == 'j') {  // Turn left
                    send_command("turn_left");
                }
                else if (key == 'm') {  // Move backward
                    send_command("move_backward");
                }
                else if (key == 'l') {  // Turn right
                    send_command("turn_right");
                }
                else if (key == 'k') {  // Stop the robot
                    send_command("stop");
                }
                else if (key == 'q') {  // Quit (exit) the program
                    RCLCPP_INFO(this->get_logger(), "Exiting control loop.");
                    stop_requested = true;
                }
            }
        } catch (const std::exception &e) {
            RCLCPP_ERROR(this->get_logger(), "Error: %s", e.what());
        }

        // Correctly end the curses session after the try block
        endwin();
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TcpClientNode>());
    rclcpp::shutdown();
    return 0;
}
